### DISCUSSION TITLE
Environment variables in eas update

### DISCUSSION THREAD

**Discussion by emehmet** [expo/expo#20742] [Archived: Questions]
When I use eas build it reads environment variables from eas.json file. But when I use eas update, the environment variables are read from the .env file.
I do not understand why?

**Comment by Ryanjso** [comment]
I ran into this issue too. Have you found a way to get eas update to use the env variables from eas.json?

**Comment by emehmet** [comment]
No, I haven't find a solution. When updating, I edit the values in the .env file as I want.

**Comment by glisom** [comment]
When running update, we pass them in as environment vars on the command like `SERVER_URL='https://...' APP_VARIANT='stage' eas update -p all --branch stage --json --auto --non-interactive`. eas.json doesn't get used during update, so we pass them manually in then use app.config.js to set like in https://docs.expo.dev/eas-update/environment-variables/

**Comment by mogulist** [comment]
I also ran into this issue. Since eas build does not use environment variables in .env, I also do not want eas update to read env vars from .env. Once unexpected env vars are injected to expo app due to new update, i could not remove unwanted env vars again via deleting specific update. I had to eas build again. Very annoying.

**Reply by mogulist** [reply]
I somehow solved this issue. After renaming `.env` to `.eeenv`, as if `.env` does not exist, I run `eas-cli build`. Once this build is completed, existence of `.env` does not matter anymore when `eas-cli update`. And i can use some environment variables in .env which are useful for local development.

I added some scripts in pakcages.json not to forget to renaming .env before build:
```
    "build:dev:local": "yarn rename:env && npx eas-cli build --profile dev -p android --local",
    "build:dev": "yarn rename:env && npx eas-cli build --profile dev -p android",
    "update:dev": "npx eas-cli update --branch dev",
    "rename:env": "test -f .env && mv .env .eeeeenv",
```

**Comment by moxorama** [comment]
The main problem is that secret keys are stored in expo secrets, and I don't want to give every developer access to them or share .env file across project.  Is there recommended way to keep keys after eas update?

**Reply by letelete** [reply]
I have the same issue. Have you found any solution to that?

**Comment by glisom** [comment]
This is our setup today, I think moving it to typescript would be the only major change we do make at this point. https://grantisom.com/2023/02/01/expo-app-config.html Downsides are losing auto-incrementing build numbers and having to do some Sentry things manually.

**Reply by letelete** [reply]
@glisom I like the setup! How do you approach setting the app's [scheme](https://docs.expo.dev/versions/latest/config/app/#scheme) with the dynamic `bundleId`? Do you set a scheme to `getBundleId` value? Or is it a union of all possible bundle IDs?

**Reply by glisom** [reply]
Just setting this to `getBundleId()`, our app doesn't have any deep linking flows as of now so that has been sufficient.

**Comment by mrkpatchaa** [comment]
I just run into this issue and thought expo-updates was not working. Hopefully crashlytics helped me find out that some variables were missing.
This is my workaround for now. 
I created a little script to create an env.xxx file and I call it before pushing update in github-actions
```
/*eslint-env node */
/* eslint-disable @typescript-eslint/no-var-requires */

const fs = require('fs-extra')

if (!process.argv[2] || ['staging', 'developmenet', 'production'].indexOf(process.argv[2]) < 0) {
  console.error('Env expected!')
  process.exit(1)
}

const easJson = require('../eas.json')

try {
  const envContent = { ...easJson.build.base.env, ...easJson.build[process.argv[2]].env }
  const envFile = Object.keys(envContent)
    .map((key) => `${key}=${envContent[key]}`)
    .join('\n')

  // console.log(envFile)
  fs.writeFileSync('.env', envFile)
} catch (error) {
  console.error(error)
}
```

```
      - name: 🚀 Push updates
        if: ${{ github.event_name == 'push' && steps.checkversion.outputs.changed == 'false' }}
        working-directory: '@mobile'
        run: |
          node scripts/touch-mobile-env.js ${{ github.ref == 'refs/heads/main' && 'production' || 'staging' }}
          eas update --auto --channel ${{ github.ref == 'refs/heads/main' && 'production' || 'staging' }}
```

**Reply by robcaldecott** [reply]
Thank you so much for this. This just fixed a borked update because of missing environment variables 🥰
